bot/cogs/advent_of_code/_cog.py

```diff
@@ -0,0 +1,103 @@
+import logging
+from datetime import datetime
+from itertools import cycle
+from operator import attrgetter
+
+import discord
+from discord import Color, Embed, app_commands
+from discord.ext import commands
+
+from bot.bot import Friendo
+from bot.settings import AOC_JOIN_CODE, AOC_LEADERBOARD_ID, AOC_SESSION_COOKIE
+from ._types import Leaderboard, LeaderboardMember
+
+logger = logging.getLogger("advent_of_code")
+
+
+class AdventOfCode(commands.GroupCog):
+    """Commands for Advent of Code."""
+
+    def __init__(self, bot: Friendo):
+        self.bot = bot
+        super().__init__()
+
+    async def fetch_leaderboard(self, year: int) -> Leaderboard:
+        """Get the leaderboard's state for a specified year."""
+        url = f"https://adventofcode.com/{year}/leaderboard/private/view/{AOC_LEADERBOARD_ID}.json"
+        cookies = {'session': AOC_SESSION_COOKIE}
+
+        # AoC Author has requested applications to provide a url to the tool in the User-Agent
+        headers = {'User-Agent': 'github.com/fisher60/friendo-bot'}
+
+        async with self.bot.session.get(url, cookies=cookies, headers=headers) as response:
+            data = await response.json()
+            return Leaderboard(**data)
+
+    @staticmethod
+    def _create_leaderboard_message(members: list[LeaderboardMember], amount: int) -> str:
+        reset = "[0m"
+        red = "[1;31m"
+        green = "[1;32m"
+        yellow = "[1;33m"
+
+        get_line_color = cycle([red, green])
+
+        formatted_message = f"Here is our top {amount}\n```ansi\n"
+        for rank, member in enumerate(members[:amount], start=1):
+            formatted_message += (
+                f"{next(get_line_color)}{rank:0>2} {yellow}★{reset} {member.name} ({member.local_score})\n"
+            )
+
+        formatted_message += "```"
+
+        return formatted_message
+
+    @app_commands.command()
+    @app_commands.describe(
+        year="View the leaderboard from a specific year, defaults to the current year",
+        amount="How many members to view, defaults to 10"
+    )
+    async def leaderboard(
+            self,
+            interaction: discord.Interaction,
+            year: app_commands.Range[int, 2015, None] = None,
+            amount: app_commands.Range[int, 0, None] = 10
+    ) -> None:
+        """Get the current Advent of Code leaderboard."""
+        if year is None:
+            year = datetime.now().year
```

Comment:
We should display the most recent AOC leader board if no year is specified. I.e this default year is only valid during December, but would attempt to retrieve a leader board that does not exist from January through November.

We should use the current year if AOC has already begun this year, otherwise use the most recent (previous year) to show the most recent standings.

Reply:
Logic updated to show the latest active AoC, [see this commit](https://github.com/fisher60/friendo-bot/pull/259/commits/a89b2bd5251a2d0332a5e709461b2f91b9d5a064)